shared/ssz/hash_cache.go

```diff
@@ -0,0 +1,247 @@
+package ssz
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/karlseguin/ccache"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/prysmaticlabs/prysm/shared/bytesutil"
+	"github.com/prysmaticlabs/prysm/shared/hashutil"
+	"github.com/prysmaticlabs/prysm/shared/params"
+)
+
+var (
+	// ErrNotMerkleRoot will be returned when a cache object is not a merkle root.
+	ErrNotMerkleRoot = errors.New("object is not a merkle root")
+	// maxCacheSize is 2x of the follow distance for additional cache padding.
+	// Requests should be only accessing blocks within recent blocks within the
+	// Eth1FollowDistance.
+	maxCacheSize = params.BeaconConfig().HashCacheSize
+	// Metrics.
+	hashCacheMiss = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "hash_cache_miss",
+		Help: "The number of hash requests that aren't present in the cache.",
+	})
+	hashCacheHit = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "hash_cache_hit",
+		Help: "The number of hash requests that are present in the cache.",
+	})
+	hashCacheSize = promauto.NewGauge(prometheus.GaugeOpts{
+		Name: "hash_cache_size",
+		Help: "The number of hashes in the block cache",
+	})
+)
+
+// hashCacheS struct with one queue for looking up by hash.
+type hashCacheS struct {
+	hashCache *ccache.Cache
+}
+
+// root specifies the hash of data in a struct.
+type root struct {
+	Hash       common.Hash
+	MerkleRoot []byte
+}
+
+// hashKeyFn takes the representation as the key for a hashInfo.
+func hashKeyFn(obj interface{}) (string, error) {
```

Comment:
Unused func

Reply:
you are right. stayed there from using a different cache method.
thanks